docs/proposals/cluster-upgrades.md

```diff
@@ -0,0 +1,104 @@
+
+This is a position statement on Cluster Upgrades.  Submitted for feedback.
+
+## Assumptions
+
+Stable API types (v1, v2, etc) will to supported for fairly long times (e.g. 1 year, and probably longer).  This doc is not about that.
+
+Critical security patches will be backported to all recetnt minor releases (maybe 1 year?  TBD).  This doc is not about that.
+
+It is about "full upgrades" of the programs that make up the cluster (docker, kubelet, apiserver, controller-manager, addons, and their respective versions and flags and many options and input files).
+in order to get new features and performance improvements.
+
+User attitudes to upgrades differ.  Here are three user archetypes:
+  - "yearly upgrader": In my company, we only upgrade software once per year, because that is how we have always done it.  Unless there is a critical security fix.
+  - "eager upgrader": I want to upgrade my cluster to every new kubernetes release.  I enjoy reading the release notes for each new release of kubernetes with a cup of tea and a chocolate croissant.  Then I try out any interesting new apiserver flags before lunch.
+  - "hosted upgrader": I use a hosted Kubernetes (e.g. GKE) and trust my hosting provider to keep my cluster up to date.  I just want it to keep working without any effort.  I do not know or care what flags my apiserver has.
+We need to have a story for all these types of users.
+
+Kubernetes project wants to release often (~4 releases per year currently, and many people would like to get to 6 or 12).  In each release, these things might change:
+  - version of kubelet required by control plane
+  - version of docker required by kubelet
+  - apiserver, controller-manager, kubelet flags
+  - storage representation of API objects
+  - kernel version and base OS config requirements
+  - existence of added fields on API objects (not captured by apiVersion) depended on by controllers, addons, etc.
+
+It is practical for the Kubernetes project to test forwards and backwards compatibility for 
+
+## Approaches to Upgrades
+
+This lists some ways we could tell our users to do upgrades.
+
+### Single-minor-version supported
+
+For an upgrade of one minor version number (e.g. 1.5 to 1.6), Kubernetes project can:
+ - provide release notes that explain requirements for upgrade
+ - provide forward/backward testing
+ - scripts to auotmate upgrade (maybe).
+ - e2e tests that ensure upgrade is possible.
+
+If Kubernetes only provides "single minor version" upgrades, then:
+  - the "yearly upgrader" customer is unhappy, because she needs to do 4 or more consecutive upgrade steps (since a jump from say 1.5 to 1.9 is unlikely to work).
+  - the "eager upgrader" and "hosted upgrader" are happy with this.
+  - the kubernetes team is happy because the upgrade testing problem is tractible
+
+### Multiple-minor-version supported
+
+If kubernetes team has to support multiple minor version upgrades, then the testing matrix gets very large.  Ick.
```

Comment:
Could we provide the ability to jump under the hood (e.g. 1.5 -> 1.9 is actually 1.5 -> 1.6 -> 1.7 -> 1.8 -> 1.9)?